### ISSUE TITLE
Build error when using ExtractTextPlugin

### ISSUE THREAD

**Issue by jacowley** [OpenNTF/dojo-webpack-plugin#23]
Thanks for creating this, it's really awesome!

I have a large application that is mostly Dojo AMD modules and partly React with ES6 modules that I'm trying to build with a single webpack configuration. However I'm getting a build error when I try to extract the CSS using extract-text-webpack-plugin.

The ES6 modules import CSS and Sass like this:
```
import "./styles.css";
```

My webpack configuration looks like this (only showing part of it here):
```
module: {
    rules: [
        {
            test: /\.jsx?$/,
            loader: "babel-loader"
        },
        {
            test: /\.css$/,
            loader: ExtractTextPlugin.extract({
                fallback: "style-loader",
                use: "css-loader"
            })
        }
    ]
},
plugins: [
    new DojoWebpackPlugin({
        loaderConfig,
        locales: ["en"],
        loader: path.resolve(__dirname, "dojoLoader/dojo/dojo.js")
    }),
    // ... [other plugins are here]
    new ExtractTextPlugin({
        filename: "[name].css",
        allChunks: true
    })
```

I managed to reproduce this using your sample application by adding a CSS file, ES6 module and including the babel-loader. This is the build error:
```
ERROR in ./css/styles.css
Module build failed: ChunkRenderError: Can't locate /var/folders/v_/tz8d71393lv7fsr707gs81kc0000gn/T/tmp-26466Cucq95eqC3xG/dojo/dojo.js in compilation
    at Compilation.createChunkAssets (/Users/jason/Development/OpenNTF/dojo-webpack-plugin-sample/node_modules/webpack/lib/Compilation.js:1275:22)
    at sealPart2 (/Users/jason/Development/OpenNTF/dojo-webpack-plugin-sample/node_modules/webpack/lib/Compilation.js:629:10)
    at Compilation.applyPluginsAsyncSeries (/Users/jason/Development/OpenNTF/dojo-webpack-plugin-sample/node_modules/tapable/lib/Tapable.js:131:46)
    at Compilation.seal (/Users/jason/Development/OpenNTF/dojo-webpack-plugin-sample/node_modules/webpack/lib/Compilation.js:579:8)
    at applyPluginsParallel.err (/Users/jason/Development/OpenNTF/dojo-webpack-plugin-sample/node_modules/webpack/lib/Compiler.js:514:17)
    at /Users/jason/Development/OpenNTF/dojo-webpack-plugin-sample/node_modules/tapable/lib/Tapable.js:225:11
    at _addModuleChain (/Users/jason/Development/OpenNTF/dojo-webpack-plugin-sample/node_modules/webpack/lib/Compilation.js:481:11)
    at processModuleDependencies.err (/Users/jason/Development/OpenNTF/dojo-webpack-plugin-sample/node_modules/webpack/lib/Compilation.js:452:13)
    at _combinedTickCallback (internal/process/next_tick.js:95:7)
    at process._tickCallback (internal/process/next_tick.js:161:9)
 @ ./js/es6module.js 6:0-28
 @ ./js/LazyContentPane.js
 @ ./js/bootstrap.js
```

If it would help I can provide my version of the sample application that I used to reproduce the error. If you can provide any help or point me in the right direction that would be great thanks.

**Comment by chuckdumont** [comment]
Does the exception text you posted match the config?  The exception is saying it can't find `/var/folders/v_/tz8d71393lv7fsr707gs81kc0000gn/T/tmp-26466Cucq95eqC3xG/dojo/dojo.js`, but your config specifies the path to the dojo loader as `path.resolve(__dirname, "dojoLoader/dojo/dojo.js")`.  Seems you copied the exception text before you added the `loader` config option to dojo-webpack-plugin

**Comment by jacowley** [comment]
No, the exception is taken from my modified version of your sample application which wasn't using the prebuilt Dojo loader.

**Comment by chuckdumont** [comment]
Please provide your changes to the sample application that reproduce the problem.  You can create a pull request with the changes against the sample project.

**Comment by chuckdumont** [comment]
For the time being, you can work around this issue by adding the following to your webpack.config.js:

````javascript
module: {
  rules: [
    {
      exclude: /dojo[\\/]dojo\.js$/,
      use: {
        loader: 'babel-loader',
      }
    }
  ]
}
````

I'll look into a solution that doesn't require this extra config.

**Comment by chuckdumont** [comment]
@jacowley please try the 2.0.9-rc1 release to see if it resolves your issue.

**Comment by jacowley** [comment]
Thanks @chuckdumont I'll give it a go.

**Comment by chuckdumont** [comment]
Leaving issue open until @jacowley verifies the fix in 2.0.9-rc1.

**Comment by jacowley** [comment]
@chuckdumont It is now fixed for the sample application but unfortunately I still get the error for our real application which has a more complex webpack configuration with multiple entry points and a commons chunk.

Maybe close this issue and I'll produce another example that reproduces my problem?

Here is the config I currently have in case you can spot the problem. It is a work-in-progress:

```
const DojoWebpackPlugin = require("dojo-webpack-plugin");
const ExtractTextPlugin = require("extract-text-webpack-plugin");
const path = require("path");
const webpack = require("webpack");
const loaderConfig = require("./dojoLoader/config");

const NODE_ENV = typeof process.env.NODE_ENV === "undefined"
    ? "development"
    : process.env.NODE_ENV;

const webpackConfig = {
    context: __dirname,

    entry: {
        vendor: [
            "./dojoLoader/dojo/dojo.js", "react", "react-dom", "prop-types"
        ],
        "core-client": [
            "babel-polyfill",
            "./modern/src/entry-point", // root of the React + ES6 module code
            "./WebProject/index"   // root of the Dojo-based AMD code
        ]
    },

    output: {
        path: path.resolve(__dirname, "WebProject/lib/release"),
        filename: "[name].js",
        publicPath: "/anaplan/js/"
    },

    module: {
        rules: [
            {
                test: /\.jsx?$/,
                include: [
                    path.resolve(__dirname, "WebProject/lib/ap"),
                    path.resolve(__dirname, "modern/src"),
                    path.resolve(__dirname, "node_modules/ap-grid")
                ],
                loader: "babel-loader"
            },
            {
                test:/\.json$/,
                loader: "json-loader"
            },
            {
                test: /\.(jpe?g|gif|png|ico|eot|ttf|svg|woff2?)$/,
                loader: "file-loader?name=[name].[ext]"
            },
            {
                test: /\.(scss|css)$/,
                loader: ExtractTextPlugin.extract({
                    fallback: "style-loader",
                    use: [
                        "css-loader",
                        "sass-loader"
                    ]
                })
            },
            {
                test: /\.mustache$/,
                loader: "raw-loader"
            }
        ]
    },

    plugins: [
        new webpack.optimize.CommonsChunkPlugin({
            name: "vendor",
            minChunks: Infinity
        }),

        new DojoWebpackPlugin({
            loaderConfig,
            locales: ["en"],
            loader: path.resolve(__dirname, "./dojoLoader/dojo/dojo.js")
        }),

        new webpack.NormalModuleReplacementPlugin(/^dojo\/query!/, function(data) {
            data.request = data.request.replace(/^dojo\/query!/, "dojo/selector/_loader!");
        }),

        new webpack.NormalModuleReplacementPlugin(/^xstyle\/css!/, function(data) {
            data.request = data.request.replace(/^xstyle\/css!/, "!css-loader!");
        }),

        new webpack.DefinePlugin({
            "process.env.NODE_ENV": `"${NODE_ENV}"`
        }),

        new ExtractTextPlugin({
            filename: "[name].css",
            allChunks: true
        }),

        new webpack.PrefetchPlugin("react"),
        new webpack.PrefetchPlugin("react-dom"),
        new webpack.NoEmitOnErrorsPlugin()
    ],

    resolve: {
        modules: [
            path.resolve(__dirname, "WebProject/lib"),
            path.resolve(__dirname, "modern/src"),
            "node_modules"
        ]
    },

    resolveLoader: {
        modules: ["node_modules"],
        extensions: [".js", ".json"],
        mainFields: ["loader", "main"]
    },

    node: {
        process: false,
        global: false
    }
}

if (NODE_ENV === "production") {
    webpackConfig.plugins.push(new webpack.optimize.UglifyJsPlugin({
        compress: {
            drop_console: true,
            warnings: false
        }
    }));
}

module.exports = webpackConfig;
```

**Comment by chuckdumont** [comment]
Just curious why you're explicitly including dojo.js in the vendor bundle.

**Comment by jacowley** [comment]
That was to work around some earlier runtime errors. So if I remove dojo from the vendor bundle I get:

```
vendor.js:346 Uncaught TypeError: Cannot read property 'call' of undefined
    at __webpack_require__ (vendor.js:346)
    at vendor.js:451
    at vendor.js:461

vendor.js:37 Uncaught TypeError: Cannot read property 'toAbsMid' of undefined
    at toAbsMid (https://local.foo.com/bar/js/vendor.js:37:40)
    at https://local.foo.com/bar/js/vendor.js:72:22
    at Array.map (native)
    at findModule (https://local.foo.com/foo/js/vendor.js:71:35)
    at contextRequire (https://local.foo.com/foo/js/vendor.js:87:22)
    at req (https://local.foo.com/foo/js/vendor.js:63:19)
    at Object.resolveTernaryHasExpression [as h] (https://local.foo.com/foo/js/vendor.js:121:24)
    at Object.id (https://local.foo.com/foo/js/core-client.js:49706:277)
    at __webpack_require__ (https://local.foo.com/foo/js/vendor.js:346:30)
    at module.exports (https://local.foo.com/foo/js/core-client.js:24925:15)
```

**Comment by chuckdumont** [comment]
So those look like runtime errors on the client.  Are you saying that when you remove dojo from the vendor bundle, you don't get the build errors anymore?

**Comment by jacowley** [comment]
Yes, the build errors go away if I remove dojo from vendor and are replaced by a runtime error. Looks like the webpack preamble is expecting to find loaderScope.require within the vendor bundle but it is undefined.

**Comment by chuckdumont** [comment]
Can you please provide the unminified vendor bundle so I can examine it?  Try zipping it and attaching it to the post.  Hopefully, it won't be too large.

**Comment by jacowley** [comment]
Here's the vendor bundle generated without dojo.js included.
[vendor.js.zip](https://github.com/OpenNTF/dojo-webpack-plugin/files/1159345/vendor.js.zip)

**Comment by chuckdumont** [comment]
OK, I think #25 will fix it.  Please try building again with 2.0.9-rc2

**Comment by jacowley** [comment]
Without Dojo in the vendor bundle it builds without any errors, but at runtime I get
```
Uncaught TypeError: webpackJsonp.registerAbsMids is not a function
    at core-client.js:sourcemap:1
```

Which is the very first line in core-client.js:
`webpackJsonp.registerAbsMids({`

I've search through both the vendor.js and core-client.js bundles and neither of them defines that function.

**Comment by chuckdumont** [comment]
Ah, I understand what's going on now.  Will have a fix shortly.

**Comment by chuckdumont** [comment]
Please try 2.0.9-rc3.  I think that will do it this time.

**Comment by jacowley** [comment]
I'm back to this error again:
```
/Users/jason/Development/core-client/core-client/node_modules/dojo-webpack-plugin/lib/DojoLoaderEnsurePlugin.js:37
                                throw Error("Can't locate " + this.options.loader + " in compilation");
                                ^

Error: Can't locate /Users/jason/Development/core-client/core-client/dojoLoader/dojo/dojo.js in compilation
    at Compilation.compilation.plugin (/Users/jason/Development/core-client/core-client/node_modules/dojo-webpack-plugin/lib/DojoLoaderEnsurePlugin.js:37:11)
    at Compilation.applyPlugins1 (/Users/jason/Development/core-client/core-client/node_modules/tapable/lib/Tapable.js:75:14)
    at Compilation.seal (/Users/jason/Development/core-client/core-client/node_modules/webpack/lib/Compilation.js:577:8)
    at applyPluginsParallel.err (/Users/jason/Development/core-client/core-client/node_modules/webpack/lib/Compiler.js:514:17)
    at /Users/jason/Development/core-client/core-client/node_modules/tapable/lib/Tapable.js:225:11
    at _addModuleChain (/Users/jason/Development/core-client/core-client/node_modules/webpack/lib/Compilation.js:481:11)
    at processModuleDependencies.err (/Users/jason/Development/core-client/core-client/node_modules/webpack/lib/Compilation.js:452:13)
    at _combinedTickCallback (internal/process/next_tick.js:131:7)
    at process._tickCallback (internal/process/next_tick.js:180:9)
```

Except now I get it even when dojo.js is not in the vendor bundle.

**Comment by jacowley** [comment]
When the error is thrown here:
[DojoLoaderEnsurePlugin.js#L33-L38](https://github.com/OpenNTF/dojo-webpack-plugin/blob/master/lib/DojoLoaderEnsurePlugin.js#L33-L38)

_compilation.modules_ has a length of 2, where:

_compilation.modules[0].rawRequest_ is:
`"!!/Users/jason/Development/core-client/core-client/node_modules/css-loader/index.js!/Users/jason/Development/core-client/core-client/node_modules/sass-loader/lib/loader.js!/Users/jason/Development/core-client/core-client/node_modules/ap-grid/src/css/Grid.css"`

_compilation.modules[1].rawRequest_ is:
`"../../../css-loader/lib/css-base.js"`

_this.options.loader_ is:
`"/Users/jason/Development/core-client/core-client/dojoLoader/dojo/dojo.js"`

Hope that helps! Let me know if there's anything I can do.

**Comment by chuckdumont** [comment]
I managed to reproduce your scenario by using the CommonChunkPlugin to extract the es6Module to a vendor bundle.  At least, I was seeing the same error you're seeing.

I think I have it sorted out now.  The issue was making sure that the Dojo loader was added to the entry chunks even when no Dojo modules are included in the compilations.  The CommonChunkPlugin and the ExtractTextPlugin both created their own challenges in this regard.  Version 2.0.9-rc4 should do the trick.  Please give it a try.

**Comment by jacowley** [comment]
YES!!! It's working! Thank you so much for all of that!
I just patched in your changes from the PR, but I'll retest properly when it has been merged in. I've still got some way to go to get our application working fully in webpack but that is a huge step forward. Thanks again.

**Comment by chuckdumont** [comment]
Fixed in release 2.0.9
